src/Workspaces/Core/Portable/Rename/Renamer.SyncNamespaceDocumentAction.cs

```diff
@@ -0,0 +1,114 @@
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.Rename
+{
+    public static partial class Renamer
+    {
+        /// <summary>
+        /// Action that will sync the namespace of the document to match the folders property 
+        /// of that document
+        /// </summary>
+        internal sealed class SyncNamespaceDocumentAction : RenameDocumentAction
+        {
+            private readonly AnalysisResult _analysis;
+
+            private SyncNamespaceDocumentAction(AnalysisResult analysis, ImmutableArray<ErrorResource> errors)
+                : base(errors)
+            {
+                _analysis = analysis;
+            }
+
+            public override string GetDescription(CultureInfo? culture)
+                => WorkspacesResources.ResourceManager.GetString("Sync_namespace_to_folder_structure", culture ?? WorkspacesResources.Culture)!;
+
+            internal override async Task<Solution> GetModifiedSolutionAsync(Document document, OptionSet _, CancellationToken cancellationToken)
+            {
+                var solution = document.Project.Solution;
+                var changeNamespaceService = document.GetRequiredLanguageService<IChangeNamespaceService>();
+                var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+                var syntaxRoot = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+                // Don't descend into anything other than top level declarations from the root.
+                // ChangeNamespaceService only controls top level declarations right now
+                var namespaceDeclarations = syntaxRoot.DescendantNodes(n => !syntaxFacts.IsDeclaration(n)).Where(n => syntaxFacts.IsNamespaceDeclaration(n));
+
+                foreach (var namespaceDeclaration in namespaceDeclarations)
+                {
+                    solution = await changeNamespaceService.ChangeNamespaceAsync(
+                        document,
+                        namespaceDeclaration,
+                        _analysis.TargetNamespace,
+                        cancellationToken).ConfigureAwait(false);
+
+                    document = solution.GetRequiredDocument(document.Id);
+                }
+
+                return solution;
+            }
+
+            public static Task<SyncNamespaceDocumentAction?> TryCreateAsync(Document document, IReadOnlyList<string> newFolders, CancellationToken _)
+            {
+                using var _1 = ArrayBuilder<ErrorResource>.GetInstance(out var errors);
+
+                var analysisResult = Analyze(document, newFolders);
+
+                if (analysisResult.HasValue)
+                {
+                    return Task.FromResult<SyncNamespaceDocumentAction?>(new SyncNamespaceDocumentAction(analysisResult.Value, errors.ToImmutable()));
```

Comment:
i don't see any place where anything is added to `errors`.

Reply:
Removed, and just passing `ImmutableArray.Empty` to base